cmd/metricscollector/v1beta1/kfp-metricscollector/v1/Dockerfile

```diff
@@ -0,0 +1,24 @@
+FROM python:3.10-slim
```

Comment:
@andreyvelich My concern is where we will put Dockerfile for KFP v2. So I would suggest we put Dockerfile for the KFP v1 on here.
wdyt?

Reply:
Oh, I see. Do we really need to support KFP v1 if, eventually, every Kubeflow users should migrate to KFP v2 ?

Reply:
Because KFP v1 and KFP v2 aren't compatible, I think migrating v1 to v2 is harder in production.
So I guess users need a lot of time to update the version.

Hence, supporting KFP v1 in Katib would be useful. WDYT?

Reply:
I see, in any case I still have a question ( https://github.com/kubeflow/katib/pull/2118#pullrequestreview-1536917369) why do we need separate Metrics Collector for KFP if we need to just read the logs from the metrics file ?

Reply:
Great point, I think if we could get away without a custom metrics collector this would be even better.

I am pretty sure I tested this in the beginning but failed.
As far as I remember the File metrics collector was not compatible with the metrics as required by KFP.

So the format is:
```
{
        "metrics": [
            {
                "name": "accuracy",  # The name of the metric. Visualized as the column name in the runs table.
                "numberValue": accuracy,  # The value of the metric. Must be a numeric value.
                "format": "PERCENTAGE",  # The optional format of the metric. Supported values are "RAW" (displayed in raw format) and "PERCENTAGE" (displayed in percentage format).
            },
            {
                "name": "val-accuracy",  # The name of the metric. Visualized as the column name in the runs table.
                "numberValue": val_accuracy,  # The value of the metric. Must be a numeric value.
                "format": "PERCENTAGE",  # The optional format of the metric. Supported values are "RAW" (displayed in raw format) and "PERCENTAGE" (displayed in percentage format).
            },
        ]
    }
```
usually dumped as a JSON.

Do I read the code of the file-metricscollector correctly, that this would require the JSON to be written on a single line:
https://github.com/kubeflow/katib/blame/89bd21f710fb4cd153a33c94e8892ec079cf63c8/cmd/metricscollector/v1beta1/file-metricscollector/main.go#L224-L228

as `logText` is a line: 
https://github.com/kubeflow/katib/blame/89bd21f710fb4cd153a33c94e8892ec079cf63c8/cmd/metricscollector/v1beta1/file-metricscollector/main.go#L176-L177

So maybe this could work if I make sure that the KFP metrics JSON file is dumped in a form where each metrics is written on a single line?

I now  added a more minimal example pipeline  that could be used to try this a bit more in-depth.

Reply:
Yeah, that's right @votti, JSON metric name and value should be on the same line. Is there any way to configure KFP keep to keep this metrics JSON file in the one line ?
Otherwise, we can introduce another [format to our file metrics collector](https://github.com/kubeflow/katib/blob/89bd21f710fb4cd153a33c94e8892ec079cf63c8/pkg/apis/controller/common/v1beta1/common_types.go#L195) to parse KFP metrics file.
Any other suggestions @tenzen-y ?

Reply:
@votti It's a great explanation! Yes, the JSON logs must be written on a single line.

> Is there any way to configure KFP keep to keep this metrics JSON file in the one line ?

If it's possible, we should take it this way.

> Otherwise, we can introduce another [format to our file metrics collector](https://github.com/kubeflow/katib/blob/89bd21f710fb4cd153a33c94e8892ec079cf63c8/pkg/apis/controller/common/v1beta1/common_types.go#L195) to parse KFP metrics file.

If the metrics format from KFP is always the following (it means users can not modify the format), we can add a new file format, `KubeflowPiplinesv1`.

> ```
>             {
>                 "name": "val-accuracy",  # The name of the metric. Visualized as the column name in the runs table.
>                 "numberValue": val_accuracy,  # The value of the metric. Must be a numeric value.
>                 "format": "PERCENTAGE",  # The optional format of the metric. Supported values are "RAW" (displayed in raw format) and "PERCENTAGE" (displayed in percentage format).
>             },
> ```

Reply:
Maybe, this is the metric form for the KFP v1?

https://github.com/kubeflow/pipelines/blob/1619f2c0b605f202c1cbb8488c64805ccff759c8/backend/api/v1beta1/run.proto#L283C11-L310

Reply:
For KFP v2, maybe this?

https://www.kubeflow.org/docs/components/pipelines/v1/sdk-v2/run-comparison/#scalar-metrics

https://github.com/kubeflow/pipelines/blob/55a2fb5c20011b01945c9867ddff0d39e9db1964/sdk/python/kfp/v2/components/types/artifact_types.py#L113-L131

Reply:
@james-jwu @zijianjoy Would you mind help us here ? We are trying to use KFP within Katib Trial and see if we can modify the output format for metrics file. Basically, save metrics in file in one line.

Reply:
Writing out the json on one line was not a big deal (actually it is already in this form. All that needed to be changed for the fileCollector (I think) is the metrics format regexp: `'"name": ([\w|-]+), "numberValue": ((-?\d+)(\.\d+)?)'`
Using this I tested the `fileCollector` approach again and run into an issue that this seems to break the argo pipeline with the error:
```
time="2023-07-25T20:53:32.488Z" level=info msg="Copying /tmp/outputs/mlpipeline_metrics/data from container base image layer to /tmp/argo/outputs/artifacts/mlpipeline-metrics.tgz"
time="2023-07-25T20:53:32.488Z" level=info msg="/var/run/argo/outputs/artifacts/tmp/outputs/mlpipeline_metrics/data.tgz -> /tmp/argo/outputs/artifacts/mlpipeline-metrics.tgz"
time="2023-07-25T20:53:32.488Z" level=error msg="executor error: open /var/run/argo/outputs/artifacts/tmp/outputs/mlpipeline_metrics/data.tgz: no such file or directory\n
```

Not sure why. I attached a failing experiment.
[katib-kfp-example-e2e-v1_filecollector.txt](https://github.com/kubeflow/katib/files/12165569/katib-kfp-example-e2e-v1_filecollector.txt)


Btw: If you wondering why I dont just use the Stdout collector and in addition print the metrics to the log: this is because this also [ broke the argo command: ](https://github.com/kubeflow/katib/issues/1914#issuecomment-1272648383) and needed a[ quite ugly workaround to still work](https://github.com/kubeflow/katib/issues/1914#issuecomment-1281480140).

Reply:
> Is there any reason for restricting metrics file configuration in one line?
@zijianjoy Katib metrics collector parses metrics file line by line and expects metrics name and value to be located in a single line.

@votti from the log line I can see that metrics are written to `/tmp/argo/outputs/artifacts/mlpipeline-metrics.tgz` file, isn't it ?

>Btw: If you wondering why I dont just use the Stdout collector and in addition print the metrics to the log: this is because this also broke the argo command: 

@votti Yeah, this could be an issue since we override the start command to make sure we redirect StdOut to `/var/log/katib/metrics.log` file, so Katib Metrics Collector can parse this file. Otherwise, Metrics Collector can't parse the StdOut. The main differences between StdOut and File metrics collector is that StdOut tails `/var/log/katib/metrics.log` file and prints logs.

Reply:
@metrics file:
One of the complexities kubeflow `pipeline` manages is to handle output artifacts (usually compressing them and storing them and saving them to an s3 storage). This is what seems to be broken when using the `filecollector`, as something while compressing and copying the file to `/tmp/argo/outputs/artifacts/mlpipeline-metrics.tgz` seems to go wrong.

After finding some time to look into it, I think the reason is very similar to the `stdout` collector:
The collector modifies the argo CMD/ARG in a way that I think causes these issues:

From the pod definition: Unmodified (eg when using the kubeflow custom metrics collector):
```
...
      _outputs = train_e2e(**_parsed_args)
      
    Args:

      --input-nr
      /tmp/inputs/input_nr/data
      --lr
      0.0005293023468535503
      --optimizer
      Adam
      --loss
      categorical_crossentropy
      --epochs
      3
      --batch-size
      36
      --mlpipeline-metrics
      /tmp/outputs/mlpipeline_metrics/data
```

When using the `filecollector` as metrics collector:

```
...
      _outputs = train_e2e(**_parsed_args)
       --input-nr /tmp/inputs/input_nr/data --lr 0.00021802007326291811 --optimizer Adam --loss categorical_crossentropy --epochs 3 --batch-size 53 --mlpipeline-metrics /tmp/outputs/mlpipeline_metrics/data && echo completed > /tmp/outputs/mlpipeline_metrics/$$$$.pid

```

I think this could be solved by following this proposal: https://github.com/kubeflow/katib/issues/2181
Until this is fixed, I think having a custom metrics collector that does not modify the command is a necessary workaround.

Reply:
@votti I think this also could solve this issue, isn't: https://github.com/kubeflow/katib/issues/577 ?
Basically, we can use Katib SDK to implement API for pushing metrics for Katib DB instead of using pull-based metrics collectors which require to change entrypoint.

For example:
```python
import kubeflow.katib as katib

client = katib.KatibClient()
client.report(metrics={"accuracy": 0.9, "loss": 0.01"})
```
We might need to do additional changes to Katib controller to verify that metrics were reported by user.

Reply:
@push based metrics collection: that sounds like a good potential solution!
So the `KatibClient` can infer automatically which `trial` these `metrics` are associated with?

Reply:
@votti Currently, user can get the Trial name using [`${trialSpec.Name}` template](https://www.kubeflow.org/docs/components/katib/trial-template/#use-trial-metadata-in-template) in their Trial's Pod environment vars. Then, user can run `KatibClient` API with appropriate Trial Name to insert metrics to the Katib DB.
I think, we should always add `TRIAL_NAME` env to the Trial pod since it is useful for many use-cases (e.g. for exporting trained model to S3, saving Trial metrics to DB, etc.)
WDYT @tenzen-y @johnugeorge ?